node/pkg/query/request.go

```diff
@@ -0,0 +1,472 @@
+	if err := binary.Read(reader, binary.BigEndian, &perChainQuery.ChainId); err != nil {
+		return fmt.Errorf("failed to read request chain: %w", err)
+	}
+
+	qt := uint8(0)
+	if err := binary.Read(reader, binary.BigEndian, &qt); err != nil {
+		return fmt.Errorf("failed to read request type: %w", err)
+	}
+	queryType := ChainSpecificQueryType(qt)
+
+	if err := ValidatePerChainQueryRequestType(queryType); err != nil {
+		return err
+	}
+
+	// Skip the query length.
+	var queryLength uint32
+	if err := binary.Read(reader, binary.BigEndian, &queryLength); err != nil {
+		return fmt.Errorf("failed to read query length: %w", err)
+	}
+
+	switch queryType {
+	case EthCallQueryRequestType:
+		q := EthCallQueryRequest{}
+		if err := q.UnmarshalFromReader(reader); err != nil {
+			return fmt.Errorf("failed to unmarshal eth call request: %w", err)
+		}
+		perChainQuery.Query = &q
+	default:
+		return fmt.Errorf("unsupported query type: %d", queryType)
+	}
+
+	return nil
+}
+
+// Validate does basic validation on a per chain query request.
+func (perChainQuery *PerChainQueryRequest) Validate() error {
+	str := perChainQuery.ChainId.String()
+	if _, err := vaa.ChainIDFromString(str); err != nil {
+		return fmt.Errorf("invalid chainID: %d", uint16(perChainQuery.ChainId))
+	}
+
+	if perChainQuery.Query == nil {
+		return fmt.Errorf("query is nil")
+	}
+
+	if err := ValidatePerChainQueryRequestType(perChainQuery.Query.Type()); err != nil {
+		return err
+	}
+
+	if err := perChainQuery.Query.Validate(); err != nil {
+		return fmt.Errorf("chain specific query is invalid: %w", err)
+	}
+
+	return nil
+}
+
+// Equal verifies that two query requests are equal.
+func (left *PerChainQueryRequest) Equal(right *PerChainQueryRequest) bool {
+	if left.ChainId != right.ChainId {
+		return false
+	}
+
+	if left.Query == nil && right.Query == nil {
+		return true
+	}
+
+	if left.Query == nil || right.Query == nil {
+		return false
+	}
+
+	if left.Query.Type() != right.Query.Type() {
+		return false
+	}
+
+	switch leftEcq := left.Query.(type) {
+	case *EthCallQueryRequest:
+		switch rightEcd := right.Query.(type) {
+		case *EthCallQueryRequest:
+			return leftEcq.Equal(rightEcd)
+		default:
+			panic("unsupported query type on right") // We checked this above!
+		}
+	default:
+		panic("unsupported query type on left") // We checked this above!
+	}
+}
+
+//
+// Implementation of EthCallQueryRequest, which implements the ChainSpecificQuery interface.
+//
+
+func (e *EthCallQueryRequest) Type() ChainSpecificQueryType {
+	return EthCallQueryRequestType
+}
+
+// Marshal serializes the binary representation of an EVM eth_call request.
+// This method calls Validate() and relies on it to range checks lengths, etc.
+func (ecd *EthCallQueryRequest) Marshal() ([]byte, error) {
+	if err := ecd.Validate(); err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	vaa.MustWrite(buf, binary.BigEndian, uint32(len(ecd.BlockId)))
```

Comment:
should these lengths be enforced to be < uint32 / uint8 accordingly?

Reply:
You got me with this one `Validate` does not check the length of `ecd.BlockId`.